telepresence/proxy/operation.py

```diff
@@ -0,0 +1,312 @@
+# Copyright 2020 Datawire. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from subprocess import CalledProcessError
+from typing import (
+    Any, Callable, Dict, Iterable, List, NamedTuple, Optional, Tuple
+)
+
+from telepresence import image_version
+from telepresence.cli import PortMapping
+from telepresence.runner import Runner
+
+from .deployment import get_image_name
+from .manifest import (
+    Manifest, make_k8s_list, make_new_proxy_pod_manifest, make_pod_manifest,
+    make_svc_manifest
+)
+from .remote import (
+    RemoteInfo, get_deployment, get_pod_for_deployment, get_remote_info,
+    make_remote_info_from_pod, wait_for_pod
+)
+
+ProxyIntent = NamedTuple(
+    "ProxyIntent", [
+        ("name", str),
+        ("container", str),
+        ("expose", PortMapping),
+        ("env", Dict[str, str]),
+        ("service_account", str),
+    ]
+)
+
+
+class ProxyOperation:
+    def __init__(self, intent: ProxyIntent) -> None:
+        self.intent = intent
+
+    def prepare(self, runner: Runner) -> None:
+        pass
+
+    def act(self, _: Runner) -> RemoteInfo:
+        raise NotImplementedError()
+
+
+LegacyOperation = Callable[[Runner, str, PortMapping, Dict[str, str], str],
+                           Tuple[str, Optional[str]]]
+
+
+class Legacy(ProxyOperation):
```

Comment:
I'm a bit confused by the convention of generic class names like `Legacy` or `New` vs something more descriptive. Is this a py3 convention or precedent you could point me to?

Reply:
I used generic names because these classes are not used outside of this package. I don't see there being a chance of confusion, since I don't expose them by importing them in this package's `__init__.py`. That said, perhaps this is a bit of Golang-style naming leaking into my Python code...